Add ability to retrieve all nested values with a wildcard parent key

snappieT opened the pull request:
Today, `PropertyFiltering` supports filtering:
- at the top level of an object
- at the top level of objects in a list
- for nested objects, __but only where the outer key is known in advance__

This generally does’t work if the response type is a Map with keys that
are not known in advance. This is the typical pattern for the response
to a multiget - you return each object keyed by its ID. This PR
addresses that requirement.

There is a new wildcard (`*`) that can be passed to property filtering.
This now returns just the properties inside the nested objects, plus
the key to these objects.

So for a JSON:
```javascript
{
   "bhalligan":{
      "firstname":"Brian",
      "lastname":"Halligan"
   },
   "spurcell":{
      "firstname":"Stephen",
      "lastname":"Purcell"
   }
}
```
And property filter: `*.firstname`, the result would be:
```javascript
{
   "bhalligan":{
      "firstname":"Brian"
   },
   "spurcell":{
      "firstname":"Stephen"
   }
}
```

@jhaber

jhaber commented on pom.xml:
```diff
@@ -10,7 +10,7 @@
     <groupId>com.hubspot.jackson</groupId>
     <artifactId>jackson-jaxrs-propertyfiltering</artifactId>
-    <version>0.8.2-SNAPSHOT</version>
+    <version>0.8.3-SNAPSHOT</version>
```
It doesn't look like there's any non-released code in master so I don't think we need to do a new release

snappieT replied:
I'll back out the releasing commits

jhaber commented:
Looking at the code, I don't think it would handle something like `property=*.firstname&property=bhalligan.lastname` which should return: 

```json
{
   "bhalligan":{
      "firstname":"Brian",
      "lastname":"Halligan"
   },
   "spurcell":{
      "firstname":"Stephen"
   }
}
```

snappieT commented:
@jhaber I finally got around to addressing this. You're right, I've updated it to now support that as you say. This involved 2 changes:
- After building the tree of filters, I now copy all wildcard filters to its siblings. In your example, this means that there is a filter for `*=firstname` and `bhalligan=firstname,lastname` as desired.
- When filtering, instead of looping over all potential filters and applying them to all keys in the JsonNode, I now loop over all keys in the JsonNode and find any filters. This allows falling back to the wildcard filter if there is no more-specific one.

I also added some new tests for your example, and also for nested exclusions. Let me know what you think of the updated PR.

tringuyen1401 commented on src/main/java/com/hubspot/jackson/jaxrs/PropertyFilter.java:
```diff
@@ -20,6 +21,22 @@ public PropertyFilter(Collection<String> properties) {
+    applyWildcardsToNamedProperties(filter);
+  }
+
+  private void applyWildcardsToNamedProperties(NestedPropertyFilter root) {
```
move this private method beneath the public ones?

tringuyen1401 commented on src/test/java/com/hubspot/jackson/jaxrs/AbstractIntegrationTest.java:
```diff
@@ -1,42 +1,17 @@
-    }
-  }
+public abstract class AbstractIntegrationTest extends BaseTest {
+  private static TypeReference<List<TestObject>> listType = new TypeReference<List<TestObject>>() { };
```
extra space between `{ }`. Also I think there's 👾 on line 12?

tringuyen1401 commented on src/test/java/com/hubspot/jackson/jaxrs/BaseTest.java:
```diff
@@ -0,0 +1,46 @@
+public abstract class BaseTest {
+
+  private static ObjectReader reader = new ObjectMapper().reader();
+
```
👾

tringuyen1401 commented on src/test/java/com/hubspot/jackson/jaxrs/BaseTest.java:
```diff
@@ -0,0 +1,46 @@
+  protected <T> T getObjects(TypeReference<T> typeReference, String path, String queryParamName, String... queryParams) throws IOException {
+    String urlString = "http://localhost:" + port + "/test" + path;
+    if (queryParams.length > 0) {
+      urlString += "?" + queryParamName +"=" + Strings.join(queryParams).with("&" + queryParamName + "=");
```
need space after the `+` between the first `queryParamName` and `"="` pair

tringuyen1401 commented on src/test/java/com/hubspot/jackson/jaxrs/NestedListIntegrationTest.java:
```diff
@@ -0,0 +1,41 @@
+
+public class NestedListIntegrationTest extends BaseTest {
+
+  private static TypeReference<List<TestNestedObject>> listNestedType = new TypeReference<List<TestNestedObject>>() { };
```
👾 between `{ }`

tringuyen1401 commented on src/test/java/com/hubspot/jackson/jaxrs/NestedObjectIntegrationTest.java:
```diff
@@ -0,0 +1,59 @@
+
+public class NestedObjectIntegrationTest extends BaseTest {
+
+  private static TypeReference<Map<Long, TestObject>> mapNestedType = new TypeReference<Map<Long, TestObject>>() { };
```
👾 between `{ }`

stevegutz commented on src/test/java/com/hubspot/jackson/jaxrs/AbstractIntegrationTest.java:
```diff
@@ -1,42 +1,16 @@
-    }
-  }
+public abstract class AbstractIntegrationTest extends BaseTest {
+  private static TypeReference<List<TestObject>> listType = new TypeReference<List<TestObject>>() {};
```
make `final`, call `LIST_TYPE`

snappieT replied:
👍 done

jhaber commented:
Rather than storing `*` directly in the `includedProperties`/`excludedProperties` collections, do you think it would simplify things to have a separate flag like `boolean wildcardIncluded`/`boolean wildcardExcluded`? 

Also, I see tests for `*.name` and `!*.name`, but could you also add tests for things like `a.*`, `!a.*`, `a.*.b`, `!a.*.b`?

snappieT commented:
@jhaber I've added those extra tests, and caught a bug that I fixed in the process. I also t[ried out adding the flags for the wildcards](https://github.com/HubSpot/jackson-jaxrs-propertyfiltering/compare/nested-map-properties...wildcard-flags), but it turns out to make the code more verbose, and to only help out in one of the places we are doing the filtering, but not in other places where we really want to get the child filters beneath the wildcard.

jhaber commented:
lgtm 🚢
